RandomSpecSorter occasionally crashes kotest with "Comparison method violates its general contract!"

**Which version of Kotest are you using**
5.1.0

### Problem statement
`SpecSorter` relies on the `fun compare(a: KClass<out Spec>, b: KClass<out Spec>): Int` defining a valid comparator object used in the `specs.sortedWith(comparator)` in the `.sort` function.

Observing the random behaviour of the `RandomSpecSorter` comparator I believe that there is some chance that an inconsistent comparator is created. This is the case when timsort algorithm is used to sort the items, as it may introspect the comparator more than once when it does multiple merges and sorts. 

The offending code:
https://github.com/kotest/kotest/blob/57909ec2030b5223a4e6c3956b7326f97f05ccd9/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/spec/sorters.kt#L32-L34

Stack trace:
```kotlin
org.junit.platform.commons.JUnitException: TestEngine with ID 'kotest' failed to execute tests
Caused by: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.base/java.util.TimSort.mergeLo(TimSort.java:781)
	at java.base/java.util.TimSort.mergeAt(TimSort.java:518)
	at java.base/java.util.TimSort.mergeCollapse(TimSort.java:448)
	at java.base/java.util.TimSort.sort(TimSort.java:245)
	at java.base/java.util.Arrays.sort(Arrays.java:1441)
	at kotlin.collections.ArraysKt___ArraysJvmKt.sortWith(_ArraysJvm.kt:2557)
	at kotlin.collections.CollectionsKt___CollectionsKt.sortedWith(_Collections.kt:1073)
	at io.kotest.engine.spec.SpecSorter$DefaultImpls.sort(sorters.kt:19)
	at io.kotest.engine.spec.RandomSpecSorter.sort(sorters.kt:32)
	at io.kotest.engine.spec.DefaultSpecExecutionOrderExtension.sort(ordering.kt:29)
	at io.kotest.engine.interceptors.SpecSortEngineInterceptor.intercept(SpecSortEngineInterceptor.kt:28)
	at io.kotest.engine.TestEngine$execute$execute$1$1.invokeSuspend(TestEngine.kt:68)
	at io.kotest.engine.TestEngine$execute$execute$1$1.invoke(TestEngine.kt)
	at io.kotest.engine.TestEngine$execute$execute$1$1.invoke(TestEngine.kt)
	... // etc
```

### Test to reproduce the issue
```kotlin
import io.kotest.assertions.throwables.shouldNotThrowAny
import io.kotest.core.spec.SpecRef
import io.kotest.core.spec.style.FunSpec
import io.kotest.engine.spec.RandomSpecSorter
import io.kotest.matchers.shouldBe
import kotlin.random.Random

class SpecSorterTest : FunSpec({
   context("random spec sorter") {
      test("should not throw 'Comparison method violates its general contract' with consistent ordering") {
         val seed = 2342731194744841942L
         val specRefs: List<SpecRef> = generateSequence { SpecRef.Reference(FunSpec::class) }.take(100).toList()
         shouldNotThrowAny { RandomSpecSorter(Random(seed)).sort(specRefs) }
      }
   }
})


// Caused by java.lang.AssertionError: No exception expected, but a IllegalArgumentException was thrown. at
//	java.base/java.util.TimSort.mergeLo(TimSort.java:781)
//	java.base/java.util.TimSort.mergeAt(TimSort.java:518)
//	java.base/java.util.TimSort.mergeCollapse(TimSort.java:448)
//	java.base/java.util.TimSort.sort(TimSort.java:245)

```

### Proposition

In order for us to preserve the behaviour I believe what we need to instead treat the collection randomization holistically as follows. 
```kotlin
/**
 * An implementation of [SpecExecutionOrder] which will order specs randomly.
 */
class RandomSpecSorter(private val random: Random) : SpecSorter {
   override fun compare(a: KClass<out Spec>, b: KClass<out Spec>): Int = 0
   override fun sort(specs: List<SpecRef>): List<SpecRef> = specs.shuffled(random) // this
}
```

This is not ideal as the compare method loses its ability to return de-stabilized comparison. However, I still think this is better than crashing the whole test runner.

--- a/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/spec/sorters.kt
+++ b/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/spec/sorters.kt
@@ -30,7 +30,8 @@ object LexicographicSpecSorter : SpecSorter {
  * An implementation of [SpecExecutionOrder] which will order specs randomly.
  */
 class RandomSpecSorter(private val random: Random) : SpecSorter {
-   override fun compare(a: KClass<out Spec>, b: KClass<out Spec>): Int = random.nextInt()
+   override fun compare(a: KClass<out Spec>, b: KClass<out Spec>): Int = 0
+   override fun sort(specs: List<SpecRef>): List<SpecRef> = specs.shuffled(random)
 }
 
 /**
@@ -41,7 +42,7 @@ class RandomSpecSorter(private val random: Random) : SpecSorter {
  * Note: Runtime annotations are not supported on Native or JS so on those platforms
  * this sort order is a no-op.
  */
-expect object AnnotatedSpecSorter :SpecSorter
+expect object AnnotatedSpecSorter : SpecSorter
 
 /**
  * An implementation of [SpecExecutionOrder] which will order specs that failed on the last run,

--- a/kotest-framework/kotest-framework-engine/src/commonTest/kotlin/com/sksamuel/kotest/engine/spec/SpecSorterTest.kt
+++ b/kotest-framework/kotest-framework-engine/src/commonTest/kotlin/com/sksamuel/kotest/engine/spec/SpecSorterTest.kt
@@ -1,0 +1,19 @@
+package com.sksamuel.kotest.engine.spec
+
+import io.kotest.assertions.throwables.shouldNotThrowAny
+import io.kotest.core.spec.SpecRef
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.engine.spec.RandomSpecSorter
+import io.kotest.matchers.shouldBe
+import kotlin.random.Random
+
+class SpecSorterTest : FunSpec({
+   context("random spec sorter") {
+      test("should not throw 'Comparison method violates its general contract' with consistent ordering") {
+         val seed = 2342731194744841942L
+         val specRefs: List<SpecRef> = generateSequence { SpecRef.Reference(FunSpec::class) }.take(100).toList()
+         val ordered = shouldNotThrowAny { RandomSpecSorter(Random(seed)).sort(specRefs) }
+         ordered shouldBe specRefs.shuffled(Random(seed))
+      }
+   }
+})

